File: tests/sdk/csharp/Schema.cs
#nullable enable

using System;
using System.Collections.Generic;

namespace Fomoxa.Net
{
    public sealed class SchemaException : Exception
    {
        public SchemaException(string message) : base(message)
        {
        }
    }

    public sealed class MessageSchema
    {
        private readonly ulong[] prefixes;

        public MessageSchema(uint id, ulong fingerprint, IReadOnlyList<ulong> prefixFingerprints)
        {
            if (prefixFingerprints == null)
            {
                throw new ArgumentNullException(nameof(prefixFingerprints));
            }
            if (prefixFingerprints.Count > ushort.MaxValue)
            {
                throw new SchemaException(
                    $"message 0x{id:X8} declares {prefixFingerprints.Count} fields, more than a u16 can carry");
            }

            var copy = new ulong[prefixFingerprints.Count];
            for (int index = 0; index < copy.Length; index++)
            {
                copy[index] = prefixFingerprints[index];
            }
            if (copy.Length > 0 && copy[copy.Length - 1] != fingerprint)
            {
                throw new SchemaException(
                    $"message 0x{id:X8}: the last prefix fingerprint is not the message fingerprint");
            }

            Id = id;
            Fingerprint = fingerprint;
            prefixes = copy;
        }

        public uint Id { get; }

        public ulong Fingerprint { get; }

        public ushort FieldCount => (ushort)prefixes.Length;

        public IReadOnlyList<ulong> PrefixFingerprints => prefixes;

        public bool TryPrefix(ushort fieldCount, out ulong fingerprint)
        {
            if (fieldCount == 0 || fieldCount > prefixes.Length)
            {
                fingerprint = 0;
                return false;
            }
            fingerprint = prefixes[fieldCount - 1];
            return true;
        }
    }

    public sealed class Schema
    {
        private readonly MessageSchema[] messages;

        public Schema(ulong fingerprint, IEnumerable<MessageSchema> messages)
        {
            if (messages == null)
            {
                throw new ArgumentNullException(nameof(messages));
            }

            var ordered = new List<MessageSchema>(messages);
            if (ordered.Count > FomoxaWire.MaxHelloMessages)
            {
                throw new SchemaException(
                    $"{ordered.Count} messages exceeds the {FomoxaWire.MaxHelloMessages} a hello may declare");
            }

            long helloSize = FomoxaWire.HelloHeaderSize + (long)FomoxaWire.HelloEntrySize * ordered.Count;
            if (helloSize > FomoxaWire.MaxHandshakePayload)
            {
                throw new SchemaException(
                    $"{ordered.Count} messages need {helloSize} hello bytes, past the " +
                    $"{FomoxaWire.MaxHandshakePayload} a handshake frame carries");
            }

            ordered.Sort((left, right) => left.Id.CompareTo(right.Id));
            for (int index = 1; index < ordered.Count; index++)
            {
                if (ordered[index - 1].Id == ordered[index].Id)
                {
                    throw new SchemaException($"message id 0x{ordered[index].Id:X8} is declared twice");
                }
            }

            Fingerprint = fingerprint;
            this.messages = ordered.ToArray();
        }

        public ulong Fingerprint { get; }

        public IReadOnlyList<MessageSchema> Messages => messages;

        public MessageSchema? Message(uint id)
        {
            int low = 0;
            int high = messages.Length;
            while (low < high)
            {
                int middle = low + ((high - low) / 2);
                if (messages[middle].Id < id)
                {
                    low = middle + 1;
                }
                else
                {
                    high = middle;
                }
            }
            return low < messages.Length && messages[low].Id == id ? messages[low] : null;
        }
    }
}

File: tests/sdk/c/src/schema.c
#include "fomoxa/schema.h"

fmx_result fmx_schema_check(const fmx_schema *schema) {
    size_t index;

    if (schema == NULL || (schema->messages == NULL && schema->message_count > 0)) {
        return FMX_ERR_INVALID;
    }
    if (schema->message_count > FMX_MAX_SCHEMA_MESSAGES) {
        return FMX_ERR_INVALID;
    }

    for (index = 0; index < schema->message_count; ++index) {
        const fmx_message_schema *message = &schema->messages[index];
        if (message->prefix_count > (size_t)UINT16_MAX) {
            return FMX_ERR_INVALID;
        }
        if (message->prefix_count > 0 && message->prefixes == NULL) {
            return FMX_ERR_INVALID;
        }
        if (message->prefix_count > 0 &&
            message->prefixes[message->prefix_count - 1] != message->fingerprint) {
            return FMX_ERR_INVALID;
        }
        if (index > 0 && schema->messages[index - 1].id >= message->id) {
            return FMX_ERR_INVALID;
        }
    }
    return FMX_OK;
}

const fmx_message_schema *fmx_schema_message(const fmx_schema *schema, uint32_t id) {
    size_t low = 0;
    size_t high;

    if (schema == NULL) {
        return NULL;
    }
    high = schema->message_count;
    while (low < high) {
        size_t middle = low + (high - low) / 2;
        if (schema->messages[middle].id < id) {
            low = middle + 1;
        } else {
            high = middle;
        }
    }
    if (low < schema->message_count && schema->messages[low].id == id) {
        return &schema->messages[low];
    }
    return NULL;
}

uint16_t fmx_message_field_count(const fmx_message_schema *message) {
    if (message == NULL) {
        return 0;
    }
    return (uint16_t)message->prefix_count;
}

bool fmx_message_prefix(const fmx_message_schema *message, uint16_t field_count, uint64_t *out) {
    if (message == NULL || field_count == 0 || (size_t)field_count > message->prefix_count) {
        return false;
    }
    *out = message->prefixes[(size_t)field_count - 1];
    return true;
}
